drivers/regulator/regulator_stm32_vrefbuf.c

```diff
@@ -0,0 +1,195 @@
+					       int32_t max_uv)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+	int32_t best_vrs = -1;
+	int32_t best_voltage = INT32_MAX;
+
+	for (uint8_t i = 0; i < config->ref_voltage_count; i++) {
+		if (!IN_RANGE(config->ref_voltages[i].uv, min_uv, max_uv) ||
+		    config->ref_voltages[i].uv >= best_voltage) {
+			continue;
+		}
+
+		best_voltage = config->ref_voltages[i].uv;
+		best_vrs = config->ref_voltages[i].vrs;
+	}
+
+	if (best_voltage == INT32_MAX) {
+		return -EINVAL;
+	}
+
+	uint8_t current_vrs = FIELD_GET(VREFBUF_CSR_VRS, config->base->CSR);
+
+	if (best_vrs == current_vrs) {
+		return 0;
+	}
+
+	config->base->CSR =
+		(config->base->CSR & ~VREFBUF_CSR_VRS) | FIELD_PREP(VREFBUF_CSR_VRS, best_vrs);
+
+	return 0;
+}
+
+static int regulator_stm32_vrefbuf_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	uint8_t vrs = FIELD_GET(VREFBUF_CSR_VRS, config->base->CSR);
+
+	for (uint8_t i = 0; i < config->ref_voltage_count; i++) {
+		if (config->ref_voltages[i].vrs == vrs) {
+			*volt_uv = config->ref_voltages[i].uv;
+			return 0;
+		}
+	}
+
+	return -EIO;
+}
+
+static int regulator_stm32_vrefbuf_init(const struct device *dev)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+	const struct device *const clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("Clock control device not ready");
+		return -ENODEV;
+	}
+
+	if (clock_control_on(clk, (clock_control_subsys_t)&config->pclken[0]) != 0) {
+		LOG_ERR("Could not enable clock");
+		return -EIO;
+	}
+
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("Reset controller not ready");
+		return -ENODEV;
+	}
+
+	if (reset_line_deassert_dt(&config->reset) != 0) {
+		LOG_ERR("Could not deassert reset line");
+		return -EIO;
+	}
+
+	if (config->vrefp_output_enable) {
+		config->base->CSR = config->base->CSR & ~VREFBUF_CSR_HIZ;
+	} else {
+		config->base->CSR = config->base->CSR | VREFBUF_CSR_HIZ;
+	}
+
+	return regulator_common_init(dev, false);
+}
+
+static DEVICE_API(regulator, api) = {
+	.enable = regulator_stm32_vrefbuf_enable,
+	.disable = regulator_stm32_vrefbuf_disable,
+	.count_voltages = regulator_stm32_vrefbuf_count_voltages,
+	.list_voltage = regulator_stm32_vrefbuf_list_voltage,
+	.set_voltage = regulator_stm32_vrefbuf_set_voltage,
+	.get_voltage = regulator_stm32_vrefbuf_get_voltage,
+};
+
+#define VREFBUF_VOLTAGE_ELEM(node_id, prop, idx)                                                   \
+	COND_CODE_1(DT_PROP_HAS_IDX(node_id, prop, UTIL_X2(idx)),                                  \
+	({		                                                                           \
+		.vrs = DT_PROP_BY_IDX(node_id, prop, UTIL_X2(idx)),		                   \
+		.uv = DT_PROP_BY_IDX(node_id, prop, UTIL_INC(UTIL_X2(idx)))	                   \
+	},), ())
```

Comment:
Step 2: update binding and instantiation macro to consume only voltage
```suggestion
#define VREFBUF_VOLTAGE_ELEM(node_id, prop, idx)                                                   \
	({		                                                                           \
		.vrs = CONCAT(LL_VREFBUF_VOLTAGE_SCALE_, idx),		                   \
		.uv = DT_PROP_BY_IDX(node_id, prop, idx)	                   \
	},
```